src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int MIN_CAPACITY = 10;
+    private static final double GROWS_COEFFICIENT = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[MIN_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (isCapacityReached()) {
+            grow();
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is out of bounds!");
+        }
+        if (isCapacityReached()) {
+            grow();
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (areElementsEquals(element, elements[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("There is no " + element + " in this ArrayList!");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        int newCapacity = (int) (elements.length * GROWS_COEFFICIENT);
+        T[] biggerArray = (T[]) new Object[newCapacity];
+        System.arraycopy(elements, 0, biggerArray, 0, elements.length);
+        elements = biggerArray;
+    }
+
+    private boolean isCapacityReached() {
+        return size == elements.length;
+    }
+
+    private void checkIndex(int index) {
```

Comment:
```suggestion
    private void validateIndex(int index) {
```
I may be wrong, but it seems to me that this name will be more informative

Reply:
There is checkIndex in Objects class that throws IndexOutOfBoundsException and that's quite similar to mine method.